### ISSUE TITLE
Pull Request ref Improvements

### ISSUE THREAD

**Issue by jakirkham** [isaacs/github#757]
For those that are unaware, there is a very nifty feature with git called remote `refs`. There are at least a few cases that one would be aware of which would be remote refs under `heads` as `refs/heads` like `refs/heads/master` for a branch called `master`. These are simply branches on the repository and would apply just as much for a local repository as a remote one. There are also all the `remotes`, which are under `refs/remotes` like `refs/remotes/origin/master` for a branch `master` on the remote `origin`. However, I digress while trying to make this accessible.

GitHub provides a nifty set of `refs` for Pull Requests that live under `pull` so `refs/pull`. They follow this up with the pull request number so `refs/pull/9` for pull request number 9. Under this there are two refs provided, which are `head` and `merge`. The `head` ref just provides the tip of the PR and `merge` is the result of merging `head` into the base branch of the PR whatever that might be (`master` or `develop` for instance). So to go back to our example one might do `get fetch origin refs/pull/9/head:pr_9` to get a branch called `pr_9` that is at the tip of PR 9. If one wants to see the merged result, they can do `get fetch origin refs/pull/9/merge:pr_9_merged` to get a branch called `pr_9_merged` that is the result of merging PR 9 with base.

This functionality is great for CIs that want to test your work or for core developers that want to review PRs without having to add remotes. There are also some neat features baked in like not constructing a `merge` ref if a PR can't be merged when opened (though there is room for improvement here). Also PRs that are closed they lose their `merge` ref. However there are two problems.

One is relatively minor, but is the lack of a `base` ref to point to the `base` branch. Admittedly one can go back and deduce this manually, but it is a tedious process and there is a chance one messes it up.

However, the second is pretty critical and could probably be more easily solved if the first one was already solved. These refs don't get updated nearly as often as they should.

For instance, if I have a PR that is able to merge it will have a `merge` ref. If the PR gets closed, the `merge` ref sensibly gets removed. However, if that PR gets merged it not only does not have its `merge` ref removed, but that `merge` ref does not actually point to the merge commit. It seems like one of these two behaviors (removal or updating) the `merge` ref for merged PRs should happen.

Another problem case includes a PR with merge conflicts. If a PR starts with merge conflicts when it is opened, it sensibly will not have a `merge` ref ( arguably because it can't ;) ). Once the conflicts are resolved a `merge` ref is reasonably added. However, if new merge conflicts arise or a working PR gets merge conflicts for the first time, the `merge` ref is not removed. Instead it remains pointing to the last point that could merge. This is a bit confusing as it suggests this is the result of the merge, but it doesn't actually point to it. Further, if the base branch of my PR gets updated and introduces merge conflicts in my PR, the `merge` ref is not updated in this case either.

In short, it would be really nice to have a `base` ref for the PRs base branch. Also it would be really nice if `merge` ref was updated as often as it needs to be.

**Comment by cirosantilli** [comment]
Recommendation: make one short request per issue, with super precise titles. More likely that things will be merged :-)

**Comment by jakirkham** [comment]
Yeah I saw them as related. If one has access to a `base` ref, it is potentially easier to determine whether to invalidate the old `merge` ref on their end and thus update it. Alternatively it would be easier to check for a merge conflict on the CI's end.

**Comment by laughedelic** [comment]
A workaround to deduce the base branch from the PR number: you can fetch the `merge` ref and because its last commit is a merge with two parents, you can do `^1` on it to get the base branch (`^2` is the PR branch): 

```
git fetch origin refs/pull/42/merge:merged_42
git checkout merged_42^1
```

**Comment by jasonk** [comment]
FYI: GitHub says `pull/#/merge` is undocumented and you shouldn't use it.

Basically they say that reason that the `merge` ref is sometimes outdated is that pushing to the PR doesn't trigger the merge to be updated, something accessing the PR from the web triggers it so that they aren't updating merges for things nobody is actually looking at.

More detail can be found at https://discourse.drone.io/t/github-claims-that-merge-refs-are-undocumented-feature/1100
